java/compilers/Symbol.java

```diff
@@ -5,26 +5,48 @@
 enum SymbolType { NONTERMINAL, TERMINAL, RULE, ALT, EOF, LAMBDA }
 
 class Symbol {
-    static final Symbol LAMBDA = new Symbol("lambda");
-    static final Symbol EOF = new Symbol("$");
+    static final Symbol LAMBDA = new Symbol("lambda", SymbolType.LAMBDA);
+    static final Symbol EOF = new Symbol("$", SymbolType.EOF);
+    private static final Symbol ALT = new Symbol("|", SymbolType.ALT);
+    private static final Symbol RULE = new Symbol("->", SymbolType.RULE);
 
     String token;
     SymbolType type;
 
-    Symbol(String token) {
+    Symbol() {
+	    token = "";
+	    type = null;
+    }
+    
+    private Symbol(String token, SymbolType type) {
         this.token = token;
-        type = matchSymbol(token);
+        this.type = type;
+    }
+
+    static Symbol of(String token) {
+        switch (token) {
+            case "->": return RULE;
+            case "|": return ALT;
+            case "$": return EOF;
+            case "lambda": return LAMBDA;
+            default:
+                if (token.matches("[a-z]+"))
```

Comment:
I wonder if using a regex is faster than just doing `token.equals(token.toUpperCase())`